### ISSUE TITLE
Dagger relies on previously-broken JavaPoet behavior that's fixed in JavaPoet 1.12.0

### ISSUE THREAD

**Issue by ZacSweers** [google/dagger#1715]
JavaPoet 1.12.0 was released this week and fixes a bug where parameter annotations were not copied over when using `MethodSpec.overriding()`. This breaks generated dagger code however, as dagger now is generating code that fails its own validation in later rounds.

### Example

given 

```kotlin
@Component
@Singleton
internal interface ExampleComponent {
  @Component.Factory
  interface Factory {
    fun create(@BindsInstance context: Context): ExampleComponent
  }
}
```

Dagger used to generate this in the resulting `DaggerExampleComponent`

```java
private static final class Factory implements ExampleComponent.Factory {
  @Override
  public ExampleComponent create(Context context) {
    Preconditions.checkNotNull(context);
    return new DaggerExampleComponent(context);
  }
}
```

However, with JavaPoet 1.12.0, it now generates this

```java
private static final class Factory implements ExampleComponent.Factory {
  @Override
  public ExampleComponent create(@BindsInstance Context context) {
    Preconditions.checkNotNull(context);
    return new DaggerExampleComponent(context);
  }
}
```

This then fails in a later processing round because Dagger reports `@BindsInstance` as not applicable to non-abstract methods

```
error: @BindsInstance parameters may only be used in abstract methods
    public ExampleComponent create(@BindsInstance Context context) {
                                                          ^
```

I believe the correct solution should be to strip dagger annotations from the created overridden methods' parameters in this case. This is possible via `toBuilder()` plus JavaPoet's new mutable builder fields, where you can just remove annotations from the builder annotations field directly.

**Comment by ronshapiro** [comment]
This reverses the behavior introduced in
https://github.com/square/javapoet/commit/9505ad0e027a1f125b5352ac722ea141831fbf1c.
It was deliberately decided that MethodSpec.overriding shouldn't copy
annotations, see https://github.com/square/javapoet/issues/482

I think we should probably revert this change from 1.12 back to what it
was.

בתאריך שבת, 11 בינו׳ 2020, 0:37, מאת Zac Sweers ‏<notifications@github.com>:

**Comment by Egorand** [comment]
Makes sense and agree with the reasoning. I'm gonna revert the change and publish 1.12.1.

**Comment by ZacSweers** [comment]
Cool, I'll close this out!

**Comment by Kolyall** [comment]
@Egorand Could not find com.squareup:javapoet:1.12.1.
```
       - https://dl.google.com/dl/android/maven2/com/squareup/javapoet/1.12.1/javapoet-1.12.1.pom
       - https://dl.google.com/dl/android/maven2/com/squareup/javapoet/1.12.1/javapoet-1.12.1.jar
       - https://jcenter.bintray.com/com/squareup/javapoet/1.12.1/javapoet-1.12.1.pom
       - https://jcenter.bintray.com/com/squareup/javapoet/1.12.1/javapoet-1.12.1.jar
       - https://repo.maven.apache.org/maven2/com/squareup/javapoet/1.12.1/javapoet-1.12.1.pom
       - https://repo.maven.apache.org/maven2/com/squareup/javapoet/1.12.1/javapoet-1.12.1.jar
       - https://jitpack.io/com/squareup/javapoet/1.12.1/javapoet-1.12.1.pom
       - https://jitpack.io/com/squareup/javapoet/1.12.1/javapoet-1.12.1.jar
```

**Comment by Egorand** [comment]
I haven't published 1.12.1 yet but will do shortly. There's a 1.13.0-SNAPSHOT that you can use meanwhile, it's on the snapshots repo.
